### ISSUE TITLE
Please add an option to include grouping operators in output

### ISSUE THREAD

**Issue by Marshallx** [jquery/esprima#1525]
The output from Esprima correctly strips grouping operators such that
`x + x + x`
gives the same AST as
`(x + x) + x`

However, it would be very useful to have an option to include a (custom) `GroupingExpression` type so that group information is not lost.

My use case is for an attempt to make a form of compiled Javascript where data type information is implied. For example:
`x + x + 1` implies addition of all three operands
but
`(x + x) + 1` indicates that the result of `x + x` cannot be inferred

The `GroupingExpression` type would have a child property `expression` which is the (only) actual JS expression wrapped. Plus the standard properties, loc, trailingComments, etc.

**Comment by ariya** [comment]
Please discuss it or bring it up at https://github.com/estree/estree. Esprima adopted former Mozilla SpiderMonkey AST format which is now standardized as ESTree. It is likely that Esprima will not diverge from the specified format for compatibility and interoperability reason.

**Comment by Marshallx** [comment]
I wasn't proposing a change to the specification of the AST (for which there would be no chance), just a way to get the grouping information. An option (off by default) wouldn't affect compatibility at all. That said, I would welcome any method of accessing grouping infromation - I just figured a custom type was the quickest and easiest way both to implement and work with.

**Comment by ariya** [comment]
Right now the best bet is to compare the node location with its parent, if the starting positions are not the same, there is likelihood that this is a grouping expression (I haven't verified all scenarios yet).

It is not easy to justify the maintenance effort for such as an optional code path to include grouping info.

**Comment by michaelficarra** [comment]
@ariya `[ 0 ]` and `[(0)]` couldn't be distinguished that way.

**Comment by Marshallx** [comment]
Hmm... `raw` only exists for Literals, doesn't it? Otherwise that would be a nice way to check - the first char of `raw` would be `(`

**Comment by michaelficarra** [comment]
@Marshallx That's not true. The node will start within the parentheses. Otherwise you could slice the original source text to look for the surrounding parentheses.

By the way, you can't look at the position *before* the node either because you wouldn't be able to tell which node was parenthesised in `(a + b)`. Maybe you could look at both sides of the node? Nope. Then non-grouping parentheses such as `if (a);` would cause false positives.

In other words, looking back at the raw source would never be the solution you're looking for.
